geometry/optimization/iris.cc

```diff
@@ -522,6 +530,495 @@ std::vector<int> revolute_joint_indices(const multibody::Joint<double>& joint) {
+  // finding a pair that is in collision and returning the corresponding index
+  int pair_in_collision = -1;
+  int i_pair = 0;
+  for (const auto& pair : sorted_pairs) {
+    plant.SetPositions(context, configuration);
+    auto query_object =
+        plant.get_geometry_query_input_port().template Eval<QueryObject<double>>(*context);
+    const double distance =
+    query_object.ComputeSignedDistancePairClosestPoints(pair.geomA, pair.geomB)
+        .distance;
+    if (distance < 0.0) {
+      pair_in_collision = i_pair;
+      break;
+    }
+    ++i_pair;
+  }
+
+  return pair_in_collision;
+}
+
+std::pair<Eigen::VectorXd, int> CollisionLineSearch(const MultibodyPlant<double>& plant, 
+const Parallelism& parallelism, const planning::CollisionChecker& checker, 
+Context<double>* context, const std::vector<GeometryPairWithDistance>& sorted_pairs, 
+const HPolyhedron& P, const Eigen::VectorXd& center, const Eigen::VectorXd& direction, 
+const double step_size = 0.05, bool parallelize_check = false, const double constraint_tol = 1e-5) {
+  // Do a line search for closest collision to center, along direction vector.
+  int i = 1;
+  int pair_in_collision = -1;
+  VectorXd collision_configuration = Eigen::VectorXd::Zero(P.ambient_dimension());
+  
+  int particle_chunk_size = parallelism.num_threads();
+  std::vector<Eigen::VectorXd> configs_to_check;
+  while (P.PointInSet(center + i * step_size * direction, constraint_tol) && pair_in_collision < 0) {
+    Eigen::VectorXd configuration;
+    for (int i_in_chunk = 0; (i_in_chunk < particle_chunk_size); ++i_in_chunk) {
+      configuration = center + i* step_size * direction;
+      configs_to_check.push_back(configuration);
+      ++i;
+      if (!P.PointInSet(center + i * step_size * direction, constraint_tol)) {
+        break;
+      }
+    }
+    std::vector<uint8_t> col_free = checker.CheckConfigsCollisionFree(configs_to_check,
+                                        parallelism);
+    bool collision_in_chunk = false;
+    for (int i_in_chunk = 0; i_in_chunk < col_free.size(); ++i_in_chunk) {
+      if (!col_free[i_in_chunk]) {
+        collision_configuration = configs_to_check[i_in_chunk];
+        collision_in_chunk = true;
+        break;
+      }
+    }
+    
+    if (collision_in_chunk) { // Have to check which pair is in collision via signed distance pair
+      pair_in_collision = FindCollisionPairIndex(plant, context, collision_configuration, sorted_pairs);
+    }
+  }
+  return std::make_pair(collision_configuration, pair_in_collision);
+}
+
+int unadaptive_test_samples(double p, double delta, double tau) {
+  return static_cast<int>(-2 * std::log(delta) / (tau * tau * p) + 0.5);
+}
+
+bool my_comparator(const VectorXd& t1, const VectorXd& t2, const Hyperellipsoid& E) {
+  return (t1 - E.center()).squaredNorm() < (t2 - E.center()).squaredNorm(); // or use a custom compare function
+}
+
+}  // namespace
+
+HPolyhedron RayIris(const MultibodyPlant<double>& plant,
```

Comment:
TODO: update interface to only take in a single const context. Note that to do this, we need SceneGraphCollisionChecker to give us not just whether or not a configuration is in collision, but if it is, which pair is in collision. Tommy will open an issue to this effect, and maybe try hacking it in himself.